File: backend/app/schemas/billing.py
from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.base import ApiModel


class VerifyPurchaseIn(ApiModel):
    purchase_token: str = Field(min_length=1)
    product_id: str | None = None
    platform: Literal["android", "ios"] = "android"
    app_account_token: str | None = None


class EntitlementOut(ApiModel):
    active: bool
    product_id: str | None = None
    state: str | None = None
    expires_at: datetime | None = None
    auto_renewing: bool = False


class FreeDesignOut(ApiModel):
    card_colors: list[str]
    card_frames: list[str]
    font_styles: list[str]
    letter_papers: list[str]


class PassInfoOut(ApiModel):
    product_id: str
    android_product_id: str
    ios_product_id: str
    title: str
    description: str
    original_price_label: str
    free_design: FreeDesignOut
    free_tones: list[str]
    free_story_lengths: list[str]
    private_postscript_requires_pass: bool

File: mobile/src/api/billing.ts
import { requestJson } from './httpClient';
import { Platform } from 'react-native';

export interface Entitlement {
  active: boolean;
  productId: string | null;
  state: string | null;
  expiresAt: string | null;
  autoRenewing: boolean;
}

export interface FreeDesign {
  cardColors: string[];
  cardFrames: string[];
  fontStyles: string[];
  imageTextures: string[];
  letterPapers: string[];
}

export interface PassInfo {
  productId: string;
  androidProductId: string;
  iosProductId: string;
  title: string;
  description: string;
  originalPriceLabel: string;
  freeDesign: FreeDesign;
  freeTones: string[];
  freeStoryLengths: string[];
  privatePostscriptRequiresPass: boolean;
}

export function fetchPassInfo() {
  return requestJson<PassInfo>('/billing/pass-info');
}

export function fetchEntitlement(token?: string | null) {
  return requestJson<Entitlement>('/billing/entitlement', { token });
}

export function verifyPurchase(
  purchaseToken: string,
  productId: string,
  platform: 'android' | 'ios',
  appAccountToken?: string,
  token?: string | null,
) {
  return requestJson<Entitlement>('/billing/verify', {
    method: 'POST',
    token,
    body: JSON.stringify({ purchaseToken, productId, platform, appAccountToken }),
  });
}

export function getPlatformPassProductId(passInfo: PassInfo | undefined): string | null {
  if (!passInfo) {
    return null;
  }
  if (Platform.OS === 'ios') {
    return passInfo.iosProductId || passInfo.productId || null;
  }
  return passInfo.androidProductId || passInfo.productId || null;
}
